File: backend/src/services/vectorStore.js
const crypto = require("crypto");
const path = require("path");
const Database = require("better-sqlite3");
const sqliteVec = require("sqlite-vec");
const { embedTexts, getProvider } = require("./llmService");

const DATABASE_FILE = path.join(__dirname, "../../data/embeddings.db");
const database = new Database(DATABASE_FILE);
sqliteVec.load(database);

database.exec(`
  CREATE TABLE IF NOT EXISTS vector_index_metadata (
    key TEXT PRIMARY KEY,
    value TEXT NOT NULL
  );
  CREATE TABLE IF NOT EXISTS document_embeddings (
    rowid INTEGER PRIMARY KEY,
    fingerprint TEXT NOT NULL,
    tab TEXT NOT NULL,
    title TEXT NOT NULL,
    text TEXT NOT NULL,
    raw TEXT NOT NULL
  );
`);

function vectorBuffer(values) {
  return Buffer.from(new Float32Array(values).buffer);
}

function getMetadata(key) {
  return database.prepare("SELECT value FROM vector_index_metadata WHERE key = ?").get(key)?.value;
}

function setMetadata(key, value) {
  database.prepare("INSERT OR REPLACE INTO vector_index_metadata (key, value) VALUES (?, ?)").run(key, value);
}

function getDocumentsFingerprint(documents) {
  return crypto.createHash("sha256")
    .update(JSON.stringify(documents.map((document) => ({ tab: document.tab, title: document.title, text: document.text, raw: document.raw }))))
    .digest("hex");
}

function recreateVectorTable(dimension) {
  database.exec("DROP TABLE IF EXISTS document_vectors");
  database.exec(`CREATE VIRTUAL TABLE document_vectors USING vec0(embedding float[${dimension}] distance_metric=cosine)`);
}

async function ensureIndex(documents) {
  const provider = getProvider();
  if (!provider || !documents.length) return false;

  const fingerprint = getDocumentsFingerprint(documents);
  const model = provider === "openai"
    ? process.env.OPENAI_EMBEDDING_MODEL || "text-embedding-3-small"
    : process.env.GEMINI_EMBEDDING_MODEL || "gemini-embedding-001";
  if (getMetadata("fingerprint") === fingerprint && getMetadata("provider") === provider && getMetadata("model") === model) {
    return true;
  }

  const vectors = await embedTexts(documents.map((document) => document.text));
  if (!vectors?.length || vectors.length !== documents.length || !vectors[0]) return false;

  recreateVectorTable(vectors[0].length);
  database.exec("DELETE FROM document_embeddings");
  const insertDocument = database.prepare("INSERT INTO document_embeddings (rowid, fingerprint, tab, title, text, raw) VALUES (?, ?, ?, ?, ?, ?)");
  const insertVector = database.prepare("INSERT INTO document_vectors (rowid, embedding) VALUES (?, ?)");
  const insertAll = database.transaction(() => {
    documents.forEach((document, index) => {
      const rowid = index + 1;
      insertDocument.run(rowid, fingerprint, document.tab, document.title, document.text, JSON.stringify(document.raw || {}));
      insertVector.run(rowid, vectorBuffer(vectors[index]));
    });
  });
  insertAll();

  setMetadata("fingerprint", fingerprint);
  setMetadata("provider", provider);
  setMetadata("model", model);
  setMetadata("dimension", String(vectors[0].length));
  return true;
}

async function search(documents, question, limit = 5) {
  if (!documents.length || !getProvider()) return null;

  try {
    const indexed = await ensureIndex(documents);
    if (!indexed) return null;
    const queryVector = (await embedTexts([question]))?.[0];
    if (!queryVector) return null;

    const matches = database.prepare(`
      SELECT document_embeddings.*, document_vectors.distance
      FROM document_vectors
      JOIN document_embeddings ON document_embeddings.rowid = document_vectors.rowid
      WHERE document_vectors.embedding MATCH ? AND k = ?
      ORDER BY document_vectors.distance
    `).all(vectorBuffer(queryVector), limit);

    return matches.map((match) => ({
      doc: {
        tab: match.tab,
        title: match.title,
        text: match.text,
        raw: JSON.parse(match.raw),
      },
      score: Math.max(0, 1 - match.distance),
    }));
  } catch (error) {
    return null;
  }
}

module.exports = { search };

File: backend-python/services/vector_store.py
from __future__ import annotations

import hashlib
import json
import sqlite3
import struct
from pathlib import Path
from typing import List, Optional

import numpy as np

from config import DATA_DIR
from services.llm_service import embed_texts, get_provider

_DB_FILE = DATA_DIR / "embeddings_py.db"


def _get_conn() -> sqlite3.Connection:
    conn = sqlite3.connect(str(_DB_FILE))
    conn.execute(
        """CREATE TABLE IF NOT EXISTS meta (key TEXT PRIMARY KEY, value TEXT NOT NULL)"""
    )
    conn.execute(
        """CREATE TABLE IF NOT EXISTS doc_embeddings (
            rowid   INTEGER PRIMARY KEY,
            tab     TEXT NOT NULL,
            title   TEXT NOT NULL,
            text    TEXT NOT NULL,
            raw     TEXT NOT NULL,
            vector  BLOB NOT NULL
        )"""
    )
    conn.commit()
    return conn


def _fingerprint(documents: List[dict]) -> str:
    payload = json.dumps(
        [{"tab": d["tab"], "title": d["title"], "text": d["text"]} for d in documents],
        sort_keys=True,
    ).encode()
    return hashlib.sha256(payload).hexdigest()


def _vec_to_blob(vec: List[float]) -> bytes:
    return struct.pack(f"{len(vec)}f", *vec)


def _blob_to_vec(blob: bytes) -> np.ndarray:
    n = len(blob) // 4
    return np.array(struct.unpack(f"{n}f", blob), dtype=np.float32)


def _cosine(a: np.ndarray, b: np.ndarray) -> float:
    na, nb = np.linalg.norm(a), np.linalg.norm(b)
    if na == 0 or nb == 0:
        return 0.0
    return float(np.dot(a, b) / (na * nb))


async def _ensure_index(documents: List[dict], conn: sqlite3.Connection) -> bool:
    if not get_provider() or not documents:
        return False

    fp = _fingerprint(documents)
    provider = get_provider()
    row = conn.execute("SELECT value FROM meta WHERE key='fingerprint'").fetchone()
    prow = conn.execute("SELECT value FROM meta WHERE key='provider'").fetchone()
    if row and row[0] == fp and prow and prow[0] == provider:
        return True

    vectors = await embed_texts([d["text"] for d in documents])
    if not vectors or len(vectors) != len(documents):
        return False

    conn.execute("DELETE FROM doc_embeddings")
    for i, (doc, vec) in enumerate(zip(documents, vectors)):
        conn.execute(
            "INSERT INTO doc_embeddings (rowid, tab, title, text, raw, vector) VALUES (?,?,?,?,?,?)",
            (i + 1, doc["tab"], doc["title"], doc["text"], json.dumps(doc.get("raw", {})), _vec_to_blob(vec)),
        )
    conn.execute("INSERT OR REPLACE INTO meta VALUES ('fingerprint',?)", (fp,))
    conn.execute("INSERT OR REPLACE INTO meta VALUES ('provider',?)", (provider,))
    conn.commit()
    return True


async def search(
    documents: List[dict], question: str, limit: int = 5
) -> Optional[List[dict]]:
    if not documents or not get_provider():
        return None

    try:
        conn = _get_conn()
        indexed = await _ensure_index(documents, conn)
        if not indexed:
            return None

        query_vecs = await embed_texts([question])
        if not query_vecs or not query_vecs[0]:
            return None
        q_vec = np.array(query_vecs[0], dtype=np.float32)

        rows = conn.execute(
            "SELECT tab, title, text, raw, vector FROM doc_embeddings"
        ).fetchall()

        scored = []
        for tab, title, text, raw_json, blob in rows:
            d_vec = _blob_to_vec(blob)
            score = _cosine(q_vec, d_vec)
            scored.append(
                (score, {"tab": tab, "title": title, "text": text, "raw": json.loads(raw_json)})
            )

        scored.sort(key=lambda x: x[0], reverse=True)
        return [{"doc": doc, "score": score} for score, doc in scored[:limit]]
    except Exception as exc:
        print(f"[vector_store] search error: {exc}")
        return None
